index.d.ts

```diff
@@ -0,0 +1,49 @@
+type TemplateFunction<T> = (template: TemplateStringsArray, ...values: any[]) => T;
+export type BoundTemplateFunction<T extends Element> = TemplateFunction<T>;
+export type WiredTemplateFunction = TemplateFunction<any>;
+
+export declare class Component<T = {}> {
+  handleEvent(e: Event): void;
+  html: WiredTemplateFunction;
+  svg: WiredTemplateFunction;
+  state: T;
+  readonly defaultState: T;
+  setState(state: Partial<T> | ((this: this, state: T) => Partial<T>)): void;
+}
+
+export declare function bind<T extends Element>(element: T): BoundTemplateFunction<T>;
+
+export declare function define(intent: string, callback: Function): void;
+
+export declare function wire(identity?: object | null, type?: 'html' | 'svg'): WiredTemplateFunction;
+export declare function wire(identity?: object | null, type_id?: string): WiredTemplateFunction;
+
+export declare const hyper: {
+  Component: typeof Component;
+  bind: typeof bind;
+  define: typeof define;
+  hyper: typeof hyper;
+  wire: typeof wire;
+
+  // hyper(null, 'html')`HTML`
```

Comment:
Hello, 

first, thank you for documenting all this :) 👍 That will be very helpful!

I have a question about how to _document_ things in TS. Guess the comments with simple double slash `//` are considered simple comments and comments with `/ ** text **/` right before a declaration are considered useful comments.

Example, if I change this line to something like this:

```ts
  /** hyper(null, 'html')`HTML` **/
  (identity: null | undefined, type?: 'html' | 'svg'): WiredTemplateFunction;

```

My editor says:
![image](https://user-images.githubusercontent.com/286066/36644688-65751062-1a5e-11e8-8735-3bd7ac819153.png)

And that could be the first step on documenting each class, fields and fn. 😉 

Cheers 🍺

Reply:
that's pretty cool but I think you don't need double stars (indeed there is an extra star showing off after the comment) and apparently markdown is enabled so that backtick transforms `` `HTML` `` into code `HTML`.

I'd say for now let's just ship types, then I'll try to describe as best as I can the rest.

I don't want to bother you too much with this already awesome PR, thanks!

Reply:
When this is merged, I'd be happy to contribute to the docs as well ;)